File: August 2025/04.py
class Solution:
    def maxRectSum(self, mat):
        def kadane(arr):
            max_sum = float('-inf')
            current_sum = 0
            for val in arr:
                current_sum = max(val, current_sum + val)
                max_sum = max(max_sum, current_sum)
            return max_sum

        if not mat or not mat[0]:
            return 0

        n = len(mat)
        m = len(mat[0])
        max_sum = float('-inf')

        for left in range(m):
            temp = [0] * n
            for right in range(left, m):
                for i in range(n):
                    temp[i] += mat[i][right]
                max_sum = max(max_sum, kadane(temp))

        return max_sum

File: August 2025/04.cpp
class Solution {
  public:
    int maxRectSum(vector<vector<int>> &mat) {
        int n = mat.size();
        int m = mat[0].size();
        int maxSum = INT_MIN;

        // Traverse all column pairs
        for (int left = 0; left < m; ++left) {
            vector<int> temp(n, 0);
            for (int right = left; right < m; ++right) {
                // Collapse columns from left to right into temp
                for (int i = 0; i < n; ++i) {
                    temp[i] += mat[i][right];
                }

                // Apply Kadane’s algorithm on temp
                int currentSum = temp[0], bestSum = temp[0];
                for (int i = 1; i < n; ++i) {
                    currentSum = max(temp[i], currentSum + temp[i]);
                    bestSum = max(bestSum, currentSum);
                }
                maxSum = max(maxSum, bestSum);
            }
        }
        return maxSum;
    }
};
